api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -132,185 +163,241 @@ public DataSet saveDataSet(
+      List<DomainValuePair> domainValuePairs,
+      Map<String, QueryParameterValue> cohortParameters,
+      boolean includesAllParticipants,
+      List<ConceptSet> conceptSetsSelected,
+      String cohortQueries) {
+
+    final ImmutableMap.Builder<String, QueryJobConfiguration> resultBuilder = new ImmutableMap.Builder<>();
+    final CdrBigQuerySchemaConfig  bigQuerySchemaConfig = cdrBigQuerySchemaConfigService.getConfig();
 
-      String conceptSetQueries =
-          conceptSetsSelected.stream()
-              .filter(cs -> d == cs.getDomainEnum())
-              .filter(cs -> d != Domain.PERSON)
-              .flatMap(cs -> cs.getConceptIds().stream().map(cid -> Long.toString(cid)))
-              .collect(Collectors.joining(", "));
-      String conceptSetListQuery = " IN (" + conceptSetQueries + ")";
-
-      if (d != Domain.PERSON) {
-        Optional<DomainConceptIds> domainConceptIds =
-            bigQuerySchemaConfig.cohortTables.values().stream()
-                .filter(config -> d.toString().equals(config.domain))
-                .map(
-                    tableConfig ->
-                        new DomainConceptIds(
-                            getColumnName(tableConfig, "source"),
-                            getColumnName(tableConfig, "standard")))
-                .findFirst();
-        if (!domainConceptIds.isPresent()) {
-          throw new ServerErrorException(
-              "Couldn't find source and standard columns for domain: " + d.toString());
-        }
-        DomainConceptIds columnNames = domainConceptIds.get();
-
-        // This adds the where clauses for cohorts and concept sets.
-        query =
-            query.concat(
-                " WHERE \n("
-                    + columnNames.getStandardConceptIdColumn()
-                    + conceptSetListQuery
-                    + " OR \n"
-                    + columnNames.getSourceConceptIdColumn()
-                    + conceptSetListQuery
-                    + ")");
-        if (!includesAllParticipants) {
-          query = query.concat(" \nAND (PERSON_ID IN (" + cohortQueries + "))");
-        }
-      } else if (!includesAllParticipants) {
-        query = query.concat(" \nWHERE PERSON_ID IN (" + cohortQueries + ")");
+    for (Domain domain : domainList) {
+      if  (domain == Domain.PERSON) {
+        continue;
+      }
+      final StringBuilder queryBuilder = new StringBuilder("SELECT ");
+
+      final List<DomainValuePair> valuePairsForCurrentDomain = domainValuePairs.stream()
+          .filter(valueSet -> valueSet.getDomain() == domain)
+          .collect(Collectors.toList());
+
+      final ValuesLinkingPair valuesLinkingPair = getValueSelectsAndJoins(
+          valuePairsForCurrentDomain);
+
+      queryBuilder.append(String.join(", ", valuesLinkingPair.getSelects()))
+          .append(" ")
+          .append(formatValuesLinkingPair(valuesLinkingPair));
+      validateSelectedConceptSetsForDomain(conceptSetsSelected);
```

Comment:
We do want this to happen for Domain.PERSON

Reply:
Do you think you could write a unit tests (against master) demonstrating what is supposed to happen with PERSON? We suspect it's not currently working correctly based on the stream logic.

Reply:
I wrote a simple one against master in the controller test. Most of that class should move into the service test class, but I gave up on doing that for this PR.